lomond/session.py

```diff
@@ -319,41 +372,54 @@ def run(self,
         # Connected to the server, but not yet upgraded to websockets
         yield events.Connected(url)
 
-        ready = False
+        selector = self._selector_cls(sock)
+        log.debug('%r created', selector)
+
+        def _regular():
+            """Run regular events if websocket is ready."""
+            if self._ready:
+                _iter_events = self._regular(
+                    poll,
+                    ping_rate,
+                    ping_timeout,
+                    close_timeout
+                )
+                for event in _iter_events:
+                    yield event
+
+        def _on_event(event):
+            """Handle logic in response to an event."""
+            if event.name == 'ready':
+                self._on_ready()
+                self._ready = True
+            elif event.name == 'ping':
+                if auto_pong:
+                    self._send_pong(event)
+            elif event.name == 'pong':
+                self._on_pong(event)
+            yield event
+            for event in _regular():
+                yield event
+
         try:
             while not websocket.is_closed:
-                reads, _errors = self._select(sock, poll)
-
-                if ready:
-                    # Check for polls / pings
-                    for event in self._regular(poll,
-                                               ping_rate,
-                                               ping_timeout,
-                                               close_timeout):
-                        yield event
-
-                if reads:
-                    data = self._recv(4096)
-                    if not data:
-                        if websocket.is_closed:
-                            break
-                        else:
+                readable = selector.wait_readable(poll)
+
+                for event in _regular():
+                    yield event
+
+                if readable:
+                    data = self._recv(64 * 1024)
```

Comment:
The Python socket docs have contained this advice since days of yore:

> Note For best match with hardware and network realities, the value of bufsize should be a relatively small power of 2, for example, 4096.

This doesn't appear to be true any more -- if it ever was. Setting it to 64K to hopefully receive our largest packets in a single chop.  autobahn test suite showed a modest speed up for this.

Reply:
Might be worth submitting an update to pydocs?